doc/neps/nep-0031-uarray.rst

```diff
@@ -0,0 +1,364 @@
+GitHub workflow. There are a few reasons for this:
+
+* Faster iteration in the case of bugs or issues.
+* Faster design changes, in the case of needed functionality.
+* ``unumpy`` will work with older versions of NumPy as well.
+* The user and library author opt-in to the override process,
+  rather than breakages happening when it is least expected.
+  In simple terms, bugs in ``unumpy`` mean that ``numpy`` remains
+  unaffected.
+
+Advantanges of ``unumpy`` over other solutions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+``unumpy`` offers a number of advantanges over the approach of defining a new
+protocol for every problem encountered: Whenever there is something requiring
+an override, ``unumpy`` will be able to offer a unified API with very minor
+changes. For example:
+
+* ``ufunc`` objects can be overridden via their ``__call__``, ``reduce`` and
+  other methods.
+* Other functions can be overridden in a similar fashion.
+* ``np.asduckarray`` goes away, and becomes ``np.asarray`` with a backend set.
+* The same holds for array creation functions such as ``np.zeros``,
+  ``np.empty`` and so on.
+
+This also holds for the future: Making something overridable would require only
+minor changes to ``unumpy``.
+
+Another promise ``unumpy`` holds is one of default implementations. Default
+implementations can be provided for any multimethod, in terms of others. This
+allows one to override a large part of the NumPy API by defining only a small
+part of it. This is to ease the creation of new duck-arrays, by providing
+default implementations of many functions that can be easily expressed in
+terms of others, as well as a repository of utility functions that help in the
+implementation of duck-arrays that most duck-arrays would require.
+
+The last benefit is a clear way to coerce to a given backend (via the
+``coerce`` keyword in ``ua.set_backend``), and a protocol
+for coercing not only arrays, but also ``dtype`` objects and ``ufunc`` objects
+with similar ones from other libraries. This is due to the existence of actual,
+third party dtype packages, and their desire to blend into the NumPy ecosystem
+(see `[6]`_). This is a separate issue compared to the C-level dtype redesign
+proposed in `[7]`_, it's about allowing third-party dtype implementations to
+work with NumPy, much like third-party array implementations. These can provide
+features such as, for example, units, jagged arrays or other such features that
+are outside the scope of NumPy.
+
+Mixing NumPy and ``unumpy`` in the same file
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Normally, one would only want to import only one of ``unumpy`` or ``numpy``,
+you would import it as ``np`` for familiarity. However, there may be situations
+where one wishes to mix NumPy and the overrides, and there are a few ways to do
+this, depending on the user's style::
+
+    import numpy.overridable as unumpy
+    import numpy as np
```

Comment:
What's the difference from simply ``import unumpy``? Also, I think it's very confusing that ``import numpy.overridable`` is an alias to a different top-level package with a completely different name.

Reply:
What we're proposing here is that a completed version of `unumpy` be vendored into NumPy as `numpy.overridable`, with a soft dependency on `unumpy` in case it exists.

Reply:
then it should be explicit where this comes from and not a top-level import: `from numpy.uarray import overridable as unumpy` or even better `from numpy.uarray import unumpy`.

Reply:
My reasoning was that people reach for `np` whenever using NumPy, so importing it as `np` keeps familiarity as the API is mirrored.

Reply:
I've changed the import to `from numpy import overridable as unumpy`.

Reply:
> then it should be explicit where this comes from and not a top-level import: `from numpy.uarray import overridable as unumpy` or even better `from numpy.uarray import unumpy`.

`unumpy` isn't a submodule of `uarray`, it's a separate module dependent on `uarray`.

Reply:
I have difficulties really judging the technical aspects of this (which is fine, I just need to spend some real time on it, but it is probably better for developers familiar with array-like libraries to judge).

My main concern is whether `numpy.overridable` is a good idea, although maybe I am confused about `unumpy` vs. `uarray` (likely my error for not reading carefully enough).  I.e. we have to put definitions somewhere so that `uarray` can use `numpy` as a backend. If we endorse `uarray` (and I think this is what this NEP is really about) putting those definitions into numpy makes sense. What I am not clear about is why that actually changes the user facing API surface?

Reply:
> I have difficulties really judging the technical aspects of this (which is fine, I just need to spend some real time on it, but it is probably better for developers familiar with array-like libraries to judge).

Feel free to make more concrete which parts are unclear and I'll add more detail.

> My main concern is whether `numpy.overridable` is a good idea, although maybe I am confused about `unumpy` vs. `uarray` (likely my error for not reading carefully enough).

`uarray` is a general override mechanism; `unumpy` is the set of definitions you speak of.

> i.e. [sic] we have to put definitions somewhere so that `uarray` can use `numpy` as a backend. If we endorse `uarray` (and I think this is what this NEP is really about) putting those definitions into numpy makes sense. What I am not clear about is why that actually changes the user facing API surface?

On acceptance, `numpy.overridable` (the name can be discussed) becomes an alias to `unumpy` if available and the vendored version if not. I guess this can be made more concrete. The reason is, this way (a separate submodule), the override is opt-in for the user rather than being some magic on the function itself. It's a partial mirror of the NumPy API.

Reply:
@seberg since we were aiming for the functionality to be opt-in (avoids some overhead and is safer, see e.g. the first reply from Nathaniel on the mailing list), there has to be a new part to the API somewhere right?

One could do this with a global switch to change the `np.<funcname>` behavior, but that's very ugly. So an explicit import with matching API seems cleanest. And there's only one thing to learn, a new import. So less mental overhead than when we're adding new functions like `asduckarray` etc.

Reply:
Yes, there has to be an opt-in/new API somewhere, but what are the real advantages of having it inside numpy proper? As far as I can see:
  1. `uarray` (really more general library than "array", which is fine) is the toolchain, a `unumpy` may want to vendor it to ensure stable API.
       - If the proposal is to try import the non-vendored version, I am actually a bit less sure about that.
  2. `unumpy` a reduced numpy-like API, which provides backend switching possibilities based on `uarray`.

An explicit import with matching API is of course the cleanest (and only reasonable thing). What I am missing is an argument why it should live within the numpy namespace or even within numpy at all and not just in the numpy organization (with documentation pointing it out).

Also maybe to be clear. Unlike `scipy.fft` the main target of `unumpy` I think are downstream packages such as `scikit-learn` or so, to allow them to easier support multiple array-likes with a single code base. These already have a big number of dependencies, so it is worth vendoring in the numpy namespace?

Reply:
Oh, there is actually:
  3. The definitions for the `numpy` backend to `unumpy` (which maybe has to live within numpy?)

Reply:
> Oh, there is actually:
> 3. The definitions for the numpy backend to unumpy (which maybe has to live within numpy?)

The backend can live as a separate package or inside `unumpy` as well.

Reply:
> These already have a big number of dependencies, so it is worth vendoring in the numpy namespace?

That's fair enough, perhaps it would be better to not vendor it. We can decide later. My main concern was that this NEP would become a discussion about adding a dependency to basically the whole scientific Python stack. Rather than about "a unified unumpy approach vs. adding more and more one-off protocols within numpy".

So, I'd like to have the latter discussion, and postpone the discussion on vendoring vs depending on.

Reply:
Sure, fair. Whether or not this can live in numpy, is not really the big deal about it, I agree. You are pointing out one thing that is currently completely missing IMO. The NEP reads as if numpy will keep spawning the need of overriding protocols. It might be worth noting that agreeing on one protocol/library (for the most part) means that it is OK if the need for new overriding protocols is not within numpy itself, but within the ecosystem.

SciPy FFT already uses it, but I did not notice too much discussion on scipy (maybe I missed it). And the use case is probably quite a lot simpler.

Reply:
In fact, if that is the main concern. Maybe the NEP should say so right in the abstract. That it is proposed that numpy provides `uarray` interface with `unumpy` and by doing so sets a precedence for allowing similar steps in the whole ecosystem (with `scipy.fft` basically anticipating the NEP).

Reply:
> SciPy FFT already uses it, but I did not notice too much discussion on scipy (maybe I missed it). And the use case is probably quite a lot simpler.

simpler in some ways, but not strictly simpler. for example, there are multiple backends there that all consume numpy arrays. a protocol like `__array_function__` that's attached to the array itself cannot deal with that.

> It might be worth noting that agreeing on one protocol/library (for the most part) means that it is OK if the need for new overriding protocols is not within numpy itself, but within the ecosystem.

I'm not sure I got your point here. can you expand on this a bit?

Reply:
> simpler in some ways, but not strictly simpler. for example, there are multiple backends there that all consume numpy arrays. a protocol like __array_function__ that's attached to the array itself cannot deal with that.

Right, it is different. I am actually very much interested in how the two would/can interact (but I already asked that a few times in the other comments I think).

> I'm not sure I got your point here. can you expand on this a bit?

Well, if Scipy uses this, and say scikit-learn uses `uarray` then you at least have the nice effect that library author can learn the same tool for many different things (and more users means better docs, etc.). Maybe not much of a point.

Reply:
Ah yes. Indeed, `uarray` is meant as a generic backend creation mechanism. There's not so many of those to choose from in Python land. Most backends are ad-hoc and specific to one library.

Reply:
> I am actually very much interested in how the two would/can interact

It'll be nested - if there's a `unumpy` function that's also covered by `__array_ufunc__` or `__array_function__`, then those protocols will be used like normal when the `numpy` backend to `unumpy` is used. In case another backend is used (dask/sparse/etc), then of course the protocols do nothing.

Reply:
> Yes, there has to be an opt-in/new API somewhere, but what are the real advantages of having it inside numpy proper? As far as I can see:
> 
>     1. `uarray` (really more general library than "array", which is fine) is the toolchain, a `unumpy` may want to vendor it to ensure stable API.
>        
>        * If the proposal is to try import the non-vendored version, I am actually a bit less sure about that.
> 
>     2. `unumpy` a reduced numpy-like API, which provides backend switching possibilities based on `uarray`.
> 
> 
> An explicit import with matching API is of course the cleanest (and only reasonable thing). What I am missing is an argument why it should live within the numpy namespace or even within numpy at all and not just in the numpy organization (with documentation pointing it out).
> 
> Also maybe to be clear. Unlike `scipy.fft` the main target of `unumpy` I think are downstream packages such as `scikit-learn` or so, to allow them to easier support multiple array-likes with a single code base. These already have a big number of dependencies, so it is worth vendoring in the numpy namespace?

While I don't mind either way, I'm +1 on moving `unumpy` into the NumPy repo and +0.7 on vendoring.